### ISSUE TITLE
Adaptive learning rate and number of iterations

### ISSUE THREAD

**Issue by huyvvo** [ppengtang/pcl.pytorch#48]
Hi, thanks for the great repo!

I tried to run the training code with both the [vgg16_voc2007](https://github.com/ppengtang/pcl.pytorch/blob/master/configs/baselines/vgg16_voc2007.yaml) and the [vgg16_voc2007_more](https://github.com/ppengtang/pcl.pytorch/blob/master/configs/baselines/vgg16_voc2007_more.yaml) configurations. I got 51.7 mAP with the former and 50.3 mAP with the latter, which are different from the reported results in [https://github.com/ppengtang/pcl.pytorch/blob/master/README.md](https://github.com/ppengtang/pcl.pytorch/blob/master/README.md) so I am trying to figure out the causes.

In [tools/train_net_step.py](https://github.com/ppengtang/pcl.pytorch/blob/master/tools/train_net_step.py), you change the learning rate at line 195:
`cfg.SOLVER.BASE_LR *= args.batch_size / original_batch_size`.
If I use `args.batch_size=1`as in the default configuration, then the base learning rate stays unchanged when I change `args.iter_size`. Since the gradients are sum-accumulated in 'inner' iterations (lines 407-421), does using large `args.iter_size` make gradients too big and non-adaptive? Should the line 195 be
`cfg.SOLVER.BASE_LR /= effective_batch_size / original_batch_size`?

Also, in lines 200-204, you change the number of iterations. I understand that it is to keep the average number of times an image is seen unchanged during training. But this changes the number of times weights are updated. Is this problematic?

**Comment by ppengtang** [comment]
Hi, since the gradients are sum-accumulated in 'inner' iterations (lines 407-421), we should use base LR 0.0005 for batch size 1 and iter size 4, and then the real LR will be base LR * iter size = 0.002. So we use `cfg.SOLVER.BASE_LR *= args.batch_size / original_batch_size` at line 195.

For lines 200-204, if we train the model for the same number of epochs and different number of batch size (or batch size * iter size), we will update the model weights fewer times. This shouldn't matter.

To reproduce the numbers, you could try different initial random seeds by changing `cfg. RNG_SEED`. In addition, in our experiments, we found that the vgg16_voc2007 configuration is more stable than vgg16_voc2007_more.

**Comment by huyvvo** [comment]
Hi, thank you for the quick response.

When we use batch size 1 and iter size 4, `cfg.SOLVER.BASE_LR *= args.batch_size / original_batch_size` and `cfg.SOLVER.BASE_LR /= effective_batch_size / original_batch_size` actually give the same base learning rate 0.0005. If we use for example batch size 1 and iter size 4, the former still gives 0.0005 while the latter gives 0.00025. The corresponding real learning rate are then 0.004 and 0.002 respectively. In this case, is 0.004 the correct real learning rate?

**Comment by ppengtang** [comment]
Sorry for the delayed reply. For batch size 1 and iter size 4, `cfg.SOLVER.BASE_LR` will be 0.0005, and the actual learning rate will be 0.0005 * iter size = 0.002. For batch size 2 and iter size 8, `cfg.SOLVER.BASE_LR` will also be 0.0005 (because `cfg.SOLVER.BASE_LR` is unrelated to iter size), and the actual learning rate will be 0.0005 * iter size = 0.004. This should be reasonable according to linear scaling rule in [this paper](https://arxiv.org/pdf/1706.02677.pdf%5B3%5D%20ImageNet). You could also tune the learning rate for iter size 8 for better results.

You could also try to tune the training iterations for better results. In my experience, for most cases we could simply half the number of iterations if we double the batch size * iter size (so we train our models the same number of epochs).

**Comment by huyvvo** [comment]
Thank you for the response. It is extremely helpful.
